misc: Change ability flyout timing to match mainline behavior

innerthunder opened the pull request:
## What are the changes the user will see?

- Ability Flyout displays are now synchronous with visual effects from ability application. In other words, the sequence of a displayed ability application is now well-defined as

```
Show flyout > Show trigger message > Apply ability > Hide flyout
```

- Async "instant" flyout displays have been fully removed. With this being removed, several abilities are no longer displayed through the ability flyout (most, if not all, of which are also not displayed in mainline)

## Why am I making these changes?

To make ability displays more stable and standardized. This also fixes ability flyout timing issues introduced in #1371.

## What are the changes from a developer perspective?

- Refactored `AbilityBar` so that the player and enemy flyout are implemented by separate `AbilityBars`. Constant values are also extracted to make ability bar positioning easier to tweak.
- Cleaned up `ShowAbilityPhase`
- Added `HideAbilityPhase` to synchronously hide either the player's or enemy's ability bar.
- Split the `apply` methods of most ability attributes into `canApply` (which checks if the effect can apply in current conditions) and `apply` (which applies the effect unconditionally)
- Some minor attribute code cleanup (e.g. replacing `Boolean/NumberHolder` instances with `ValueHolder`)

## Screenshots/Videos

https://github.com/user-attachments/assets/609c3d39-19f3-41b4-8729-bd3a97d4817d

<details><summary>Overrides</summary>
<p>

```ts
const overrides = {
  ENEMY_ABILITY_OVERRIDE: AbilityId.LEVITATE,
  ENEMY_MOVESET_OVERRIDE: MoveId.EARTHQUAKE,
  PASSIVE_ABILITY_OVERRIDE: AbilityId.LEVITATE,
  MOVESET_OVERRIDE: MoveId.EARTHQUAKE,
} satisfies Partial<InstanceType<typeof DefaultOverrides>>;
```

</p>
</details> 

## How to test the changes?

The ability flyout needs to be tested manually. Using overrides to force a post-summon ability (e.g. Intimidate) is probably the fastest and easiest way to test.

## Checklist

- [ ] ⚠️ If this is a PR for `main` (such as a hotfix), has the game version been updated (`pnpm update-version:patch` / `pnpm update-version:minor`?
- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`pnpm test:silent`)
  - [x] Have I created new automated tests (`pnpm test:create`) or updated existing tests related to the PR's changes?
- [x] Have I provided screenshots/videos of the changes (if applicable)?
  - [x] Have I made sure that any UI change works for both UI themes (dark and light)?

innerthunder committed:
Update `AbilityBar` and related phases

innerthunder committed:
Add return type to `playTween`

innerthunder committed:
Fix compile errors

innerthunder committed:
Merge branch 'beta' into refactor/show-ability-timing

innerthunder committed:
Added `canApply` to ability attributes

innerthunder committed:
Integrate `canApply` into `applyAbAttrsInternal`

innerthunder committed:
Remove `showAbilityInstant` + update attributes' `showAbility` values

innerthunder committed:
Fix issues in existing tests

innerthunder committed:
typecheck fixes

innerthunder committed:
Remove `AbilityBar` auto-hide functionality

innerthunder committed:
Re-add enemy ability bar

innerthunder committed:
Merge branch 'beta' into refactor/show-ability-timing

innerthunder committed:
Apply `canApply` split to Victory Dance / Ruin ability attributes

innerthunder committed:
Fix stalling phases

innerthunder committed:
Merge branch 'beta' into refactor/show-ability-timing

innerthunder committed:
Fix visual bugs + type immunity ability display

innerthunder committed:
Remove todo

DayKev commented on src/data/abilities/ab-attrs/pre-apply-battler-tag-immunity-ab-attr.ts:
The chain of `AbAttr`s involving this one are a bit... lol

innerthunder replied:
yep... something to untangle later 😆

DayKev commented on src/data/abilities/ab-attrs/force-switch-out-immunity-ab-attr.ts:
```diff
@@ -1,16 +1,15 @@
 export class ForceSwitchOutImmunityAbAttr extends AbAttr {
-  constructor(showAbility: boolean = true, showAbilityInstant: boolean = false) {
-    super(showAbility, showAbilityInstant);
+  constructor() {
+    super();
```
Suction Cups should show a flyout: https://www.youtube.com/watch?v=lW6p9GCrRS4
and Guard Dog shouldn't (for some reason...): https://www.youtube.com/watch?v=rggthHKGgTM
```suggestion
  constructor(showAbility: boolean) {
    super(showAbility);
```

innerthunder replied:
This would require an almost full rewrite of the attribute because the activation message is currently handled externally wherever this is applied. It used to use an async flyout alongside the external messages to show the ability.

Also, I'd rather just display Guard Dog anyway. Both abilities currently use a generic "can't be switched out" message when activated. Should they use different messages?

DayKev replied:
Having both display the flyout is fine, it makes more sense that way (Gamefreak please...).
Suction Cups has its own message, but Guard Dog can continue to use a generic one.

Tempo-anon replied:
Just slap a comment and call it a day. It's not a major thing and a lot of Pokemon fangames have qol ability flyouts even when it's not technically canon 🤷

innerthunder replied:
Added a TODO to the attribute

DayKev commented on src/data/abilities/ab-attrs/full-hp-resist-type-ab-attr.ts:
```diff
@@ -24,27 +24,22 @@ export class FullHpResistTypeAbAttr extends PreDefendAbAttr {
-    }
-    return false;
+  public override canApply(...[pokemon, , , move, typeMultiplier]: Parameters<this["apply"]>): boolean {
+    return pokemon.isFullHp() && !move.hasAttr(FixedDamageAttr) && typeMultiplier.value > 0.5;
```
Just a small micro-optimization
```suggestion
    return pokemon.isFullHp() && typeMultiplier.value > 0.5 && !move.hasAttr(FixedDamageAttr);
```

DayKev commented on src/data/abilities/ab-attrs/mock-status-effect-ab-attr.ts:
```diff
@@ -11,14 +11,13 @@ import type { NumberHolder } from "#utils/common-utils";
   public mockedStatus: StatusEffect;
   constructor(mockedStatus: StatusEffect) {
-    super(false);
+    super();
```
Hm, Comatose is supposed to have an ability flyout (once when it comes out, and whenever it blocks a status move https://www.youtube.com/watch?v=I1Eoh6OFq38) but the current setup means you can't just pass `true` here...

innerthunder replied:
Made an issue #1376

DayKev commented on src/data/abilities/ab-attrs/move-effect-chance-multiplier-ab-attr.ts:
```diff
@@ -3,41 +3,36 @@ import { AbAttrFlag } from "#enums/ab-attr-flag";
+   * Secondary effect chance multipliers do not apply to these moves
+   * even though they are implemented with 100 base chance.
+   */
+  private static exceptMoves = [MoveId.ORDER_UP, MoveId.ELECTRO_SHOT];
```
Instead of a `private` `static` class member, you could just make it a non-exported `const` outside of the class. It would minify better that way too.
```suggestion
  private static readonly exceptMoves: readonly MoveId[] = [MoveId.ORDER_UP, MoveId.ELECTRO_SHOT];
```

DayKev commented on src/data/abilities/ab-attrs/move-effect-chance-multiplier-ab-attr.ts:
```diff
@@ -3,41 +3,36 @@ import { AbAttrFlag } from "#enums/ab-attr-flag";
-
+  public override apply(_pokemon: Pokemon, _simulated: boolean, moveChance: ValueHolder<number>, _move: Move): void {
     moveChance.value *= this.chanceMultiplier;
     moveChance.value = Math.min(moveChance.value, 100);
```
```suggestion
    moveChance.value = Math.min(moveChance.value * this.chanceMultiplier, 100);
```

DayKev commented on src/data/abilities/ab-attrs/non-super-effective-immunity-ab-attr.ts:
```diff
@@ -17,20 +17,19 @@ export class NonSuperEffectiveImmunityAbAttr extends TypeImmunityAbAttr {
+    cancelled: ValueHolder<boolean>,
+    typeMultiplier: ValueHolder<number>,
+  ): void {
+    cancelled.value = true; // Suppresses "No Effect" message
```
This isn't necessary anymore since you're making it call `super.appy`, which you changed to set `cancelled.value = true`.
```suggestion
```

DayKev commented on src/data/abilities/ab-attrs/type-immunity-ab-attr.ts:
```diff
@@ -14,50 +15,54 @@ import type { BooleanHolder, NumberHolder } from "#utils/common-utils";
+    return i18next.t("battle:hitResultNoEffect", { pokemonName: getPokemonNameWithAffix(pokemon) });
+  }
+
+  public override getCondition(): AbAttrCondition {
     return this.condition;
   }
```
Hmm... there's three layers of conditions: `AbAttr#extraCondition` which is used by `.conditionalAttr`, the various `[ability attr]#condition` fields that are added on which are usually (but apparently not always) evaluated within `canApply`, and `AbAttr#canApply` itself. There doesn't appear to be a difference between whether `extraCondition` evaluates to false or `canApply` returns false either, from looking at `apply-ab-attrs.ts`. It should be possible to consolidate all of this somehow (in the future).

innerthunder replied:
Made an issue #1378

DayKev commented on src/data/abilities/ab-attrs/post-battle-loot-ab-attr.ts:
```diff
@@ -2,28 +2,42 @@ import { PostBattleAbAttr } from "#abilities/post-battle-ab-attr";
+    if (
+      !simulated
+      && globalScene.tryTransferHeldItemModifier(this.randItem, pokemon, true, 1, true, undefined, false)
+    ) {
```
```suggestion
    if (globalScene.tryTransferHeldItemModifier(this.randItem, pokemon, true, 1, true, undefined, false)) {
```

DayKev commented on src/data/abilities/ab-attrs/post-battle-loot-ab-attr.ts:
```diff
@@ -2,28 +2,42 @@ import { PostBattleAbAttr } from "#abilities/post-battle-ab-attr";
+  private randItem?: PokemonHeldItemModifier;
+
+  public override apply(pokemon: Pokemon, simulated: boolean, _isVictory: boolean): void {
     const postBattleLoot = globalScene.currentBattle.postBattleLoot;
```
Technically this `AbAttr` doesn't need to check `simulated` because it should never be `true` for this.
```suggestion
    if (simulated) {
      return;
    }
    const postBattleLoot = globalScene.currentBattle.postBattleLoot;
```

DayKev commented on src/data/abilities/ab-attrs/post-biome-change-ab-attr.ts:
```diff
@@ -1,9 +1,12 @@
 import { AbAttrFlag } from "#enums/ab-attr-flag";
+import type { Pokemon } from "#field/pokemon";
 
 export abstract class PostBiomeChangeAbAttr extends AbAttr {
```
Huh... this `AbAttr` and its subclasses are entirely reundant, `NewBiomeEncounterPhase` could just do `applyAbAttrs<PostSummonTerrainChangeAbAttr>(...)` and `applyAbAttrs<PostSummonWeatherChangeAbAttr>(...)` since they have the exact same code as these `AbAttr`s.

Though it probably doesn't even need to do that anymore, I think that was solely for Endless (since the X5 biome changes didn't cause an arena reset) which was removed. So just delete this `AbAttr` and it's subclasses, and remove the application from `NewBiomeEncounterPhase`.

innerthunder replied:
`NewBiomeEncounterPhase` is entirely unused since Endless was removed. Should it be removed as well?

DayKev replied:
Ah, yeah.

innerthunder replied:
I marked it as deprecated. Removing it here seems kind of out of scope.

DayKev commented on src/data/abilities/ab-attrs/post-defend-steal-held-item-ab-attr.ts:
```diff
@@ -1,48 +1,53 @@
+    }
+
+    const heldItems = this.getTargetHeldItems(attacker).filter((i) => i.isTransferable);
+    return heldItems.length > 0;
```
```suggestion
    return this.getTargetHeldItems(attacker).some((i) => i.isTransferable);
```

DayKev commented on src/data/abilities/ab-attrs/post-summon-copy-ability-ab-attr.ts:
```diff
@@ -14,37 +13,24 @@ export class PostSummonCopyAbilityAbAttr extends PostSummonAbAttr {
+      pokemon.summonData.ability = this.target.getAbility().id;
+      this.target.waveData.abilitiesRevealed.push(this.target.getAbility().id);
+      pokemon.updateInfo();
     }
```
Reduces `.getAbility()` calls from 3 to 1, and removes the unnecessary `allAbilities[abilityId]`.
```suggestion
    if (simulated) {
      return;
    }

    const targetAbility = this.target.getAbility();
    this.targetAbilityName = targetAbility.name;
    pokemon.summonData.ability = targetAbility.id;
    this.target.waveData.abilitiesRevealed.push(targetAbility.id);
    pokemon.updateInfo();
```

DayKev commented on src/data/abilities/ab-attrs/post-summon-user-field-remove-status-effect-ab-attr.ts:
```diff
@@ -11,34 +11,32 @@ import { getStatusEffectHealText } from "#utils/status-effect-utils";
+
+  public override canApply(...[pokemon]: Parameters<this["apply"]>): boolean {
+    const affectedPokemon = pokemon.getField().filter((p) => p.isActive(true));
+    return affectedPokemon.some((p) => p.hasStatusEffect(this.statusEffects, false, true));
```
```suggestion
    return pokemon.getField().some((p) => p.isActive(true) && p.hasStatusEffect(this.statusEffects, false, true));
```

DayKev commented on src/data/abilities/ab-attrs/anticipation-ab-attr.ts:
```diff
@@ -9,23 +10,38 @@ import { OneHitKOAttr } from "#moves/one-hit-ko-attr";
+  public override apply(pokemon: Pokemon, simulated: boolean): void {
+    super.apply(pokemon, simulated);
+  }
+
```
This doesn't appear to be necessary (anymore?).
```suggestion
```

DayKev commented on src/data/abilities/ab-attrs/prevent-bypass-speed-chance-ab-attr.ts:
```diff
@@ -15,19 +15,20 @@ export class PreventBypassSpeedChanceAbAttr extends AbAttr {
-    }
-    return false;
+
+    return move != null && this.condition(pokemon, move) && isCommandFight;
```
Another micro-optimization.
```suggestion
    return move != null && isCommandFight && this.condition(pokemon, move);
```

DayKev commented on src/data/abilities/ab-attrs/reduce-berry-use-threshold-ab-attr.ts:
```diff
@@ -1,22 +1,19 @@
-    return false;
+  public override canApply(...[pokemon, , threshold]: Parameters<this["apply"]>): boolean {
+    return isBetween(pokemon.getHpRatio(), threshold.value, threshold.value * 2);
   }
```
The Ability should always apply, since it's not what determines whether the berry activates or not, it's just changing it from `<= 25%` to `<= 50%`.
```suggestion
```

DayKev commented on src/data/abilities/ab-attrs/suppress-weather-effect-ab-attr.ts:
```diff
@@ -2,9 +2,13 @@ import { PreWeatherEffectAbAttr } from "#abilities/pre-weather-effect-ab-attr";
+   * @todo Should this be removed?
+   * Every instance of this attr sets this to `true`.
+   */
   public readonly affectsPrimal: boolean;
```
I think it's safe to remove it. If necessary it can be re-added later, though it probably won't need to be.

DayKev commented on src/field/arena.ts:
```diff
@@ -376,6 +376,14 @@ export class Arena {
+   * @returns `true` if Terrain of the given type can be set on the field.
+   */
+  public canSetTerrain(terrain: TerrainType): boolean {
+    return this.terrain?.terrainType !== terrain;
```
```suggestion
    return (this.terrain?.terrainType ?? TerrainType.NONE) !== terrain;
```

DayKev commented on src/utils/anim-utils.ts:
```diff
@@ -73,14 +73,8 @@ export async function loadEncounterAnimAssets(startLoad?: boolean): Promise<void
  * @param config - The config for a single Tween
  * @param scene - The {@linkcode SceneBase} on which the Tween plays (Default {@linkcode globalScene})
- *
- * @privateRemarks
- * The `config` input should not include an `onComplete` field as that callback is
- * used to resolve the Promise containing the Tween animation. However, `config`'s type
- * cannot be changed to something like `Omit<TweenBuilderConfig, "onComplete">` due to
- * how Phaser interprets `TweenBuilderConfig` inputs.
```
Why remove the comment?

DayKev commented on src/battle-scene.ts:
```diff
@@ -2146,6 +2152,51 @@ export class BattleScene extends SceneBase {
 
+  private hasModifier(modifier: PersistentModifier, enemy = false): boolean {
+    const modifiers = enemy ? this.enemyModifiers : this.modifiers;
+    return modifiers.indexOf(modifier) > -1;
```
```suggestion
    return modifiers.includes(modifier);
```

DayKev commented on src/battle-scene.ts:
```diff
@@ -2146,6 +2152,51 @@ export class BattleScene extends SceneBase {
+    const source = mod.pokemonId ? mod.getPokemon() : null;
+    const cancelled = new ValueHolder(false);
+
+    if (source && source.isPlayer() !== target.isPlayer()) {
```
```suggestion
    if (source?.isOpponent(target)) {
```

DayKev committed:
Update docs/comments

DayKev committed:
Remove `const [...] = params;`

DayKev committed:
Fix typos

DayKev committed:
Add some `readonly`

DayKev committed:
Fix Gluttony ability

DayKev committed:
Misc changes

- Replace an incorrectly used `.map` with `.forEach`

- Remove unnecessary `.forceHit()` in Effect Spore test

DayKev committed:
De-nesting, remove redundant `simulated` check

Remove redundant check for `simulated`
in `post-summon-copy-ally-stats-ab-attr.ts`

DayKev committed:
Convert missed `apply` methods

DayKev commented on src/data/abilities/ab-attrs/stab-boost-ab-attr.ts:
```diff
@@ -51,7 +44,11 @@ export class StabBoostAbAttr extends AbAttr {
+  public override canApply(...[pokemon, , move]: Parameters<this["apply"]>): boolean {
+    return (
+      pokemon.getMoveType(move) !== ElementalType.STELLAR && pokemon.getTypes(true).includes(pokemon.getMoveType(move))
+    );
```
1 less function call.
```suggestion
    const moveType = pokemon.getMoveType(move);
    return moveType !== ElementalType.STELLAR && pokemon.getTypes(true).includes(moveType);
```

DayKev commented on src/data/abilities/ab-attrs/synchronize-status-ab-attr.ts:
```diff
@@ -10,37 +10,36 @@ import type { Pokemon } from "#field/pokemon";
-  constructor(showAbility: boolean = true, showAbilityInstant: boolean = false) {
-    super(showAbility, showAbilityInstant);
+  /** All status effects that can be transferred by Synchronize */
+  private static syncStatuses = new Set<StatusEffect>([
```
Same comment as in `MoveEffectChanceMultiplierAbAttr` re: private static fields -> un-exported `const` outside the class.
```suggestion
  private static readonly syncStatuses = new ReadonlySet<StatusEffect>([
```

DayKev committed:
Re-add incorrectly removed constructor for Gorilla Tactics `AbAttr`

DayKev committed:
Fix Poison Puppeteer `AbAttr` constructor

innerthunder committed:
Merge branch 'beta' into refactor/show-ability-timing

innerthunder committed:
Review suggestions + minor cleanup

Co-Authored-By: NightKev <34855794+DayKev@users.noreply.github.com>

innerthunder committed:
Fix typecheck issues

innerthunder committed:
Fix infinite loop from `Weather.isEffectSuppressed`

innerthunder committed:
Fix ME test (?)

innerthunder committed:
Merge branch 'beta' into refactor/show-ability-timing

innerthunder committed:
apply linter

innerthunder committed:
Add docs + TODO for `ForceSwitchOutImmunityAbAttr`

DayKev commented on src/utils/ability-utils.ts:
```diff
@@ -34,7 +34,20 @@ export function getWeatherCondition(...weatherTypes: WeatherType[]): AbAttrCondi
+ * Used for {@link https://bulbapedia.bulbagarden.net/wiki/Pixilate_(Ability) | Pixilate} et al.
+ */
+export const normalTypeMoveConversionCondition: PokemonAttackCondition = (user, target, move) =>
+  anyTypeMoveConversionCondition(user, target, move) && move?.type === ElementalType.NORMAL;
```
```suggestion
  move?.type === ElementalType.NORMAL && anyTypeMoveConversionCondition(user, target, move);
```

DayKev commented on src/utils/ability-utils.ts:
```diff
@@ -34,7 +34,20 @@ export function getWeatherCondition(...weatherTypes: WeatherType[]): AbAttrCondi
+/**
+ * @returns `true` if the move isn't a variable-type move. Tera-based variable-type moves may
+ * still have their type changed by abilities with this condition if the user is not Terastallized.
+ *
```
```suggestion
 * @remarks
```

innerthunder committed:
Apply review suggestions

Co-authored-by: NightKev <34855794+DayKev@users.noreply.github.com>

DayKev committed:
Merge branch 'beta' into refactor/show-ability-timing

innerthunder committed:
clean up external `ShowAbilityPhase` scheduling

innerthunder committed:
Merge branch 'refactor/show-ability-timing' of https://github.com/Despair-Games/poketernity into refactor/show-ability-timing

innerthunder committed:
Simplify anti-redirection logic

DayKev committed:
Merge branch 'beta' into refactor/show-ability-timing
